contracts/staking/PoSValidatorManager.sol

```diff
@@ -130,4 +140,155 @@ abstract contract PoSValidatorManager is IPoSValidatorManager, ValidatorManager
+        PoSValidatorManagerStorage storage $ = _getPoSValidatorManagerStorage();
+        require(
+            $._pendingRegisterDelegatorMessages[validationID][delegator].length > 0
+                && $._delegatorStakes[validationID][delegator].status == ValidatorStatus.PendingAdded,
+            "PoSValidatorManager: Delegator registration not pending"
+        );
+
+        // Submit the message to the Warp precompile.
+        WARP_MESSENGER.sendWarpMessage($._pendingRegisterDelegatorMessages[validationID][delegator]);
+    }
+
+    function completeDelegatorRegistration(uint32 messageIndex, address delegator) external {
+        PoSValidatorManagerStorage storage $ = _getPoSValidatorManagerStorage();
+
+        // Unpack the Warp message
+        WarpMessage memory warpMessage = _getPChainWarpMessage(messageIndex);
+        (bytes32 validationID, uint64 nonce, uint64 weight) =
+            ValidatorMessages.unpackSubnetValidatorWeightUpdateMessage(warpMessage.payload);
+        require(
+            $._pendingRegisterDelegatorMessages[validationID][delegator].length > 0
+                && $._delegatorStakes[validationID][delegator].status == ValidatorStatus.PendingAdded,
+            "PoSValidatorManager: Delegator registration not pending"
+        );
+        delete $._pendingRegisterDelegatorMessages[validationID][delegator];
+
+        // Update the validator weight and delegator status
+        Validator memory validator = _getValidator(validationID);
+        require(
+            $._delegatorStakes[validationID][delegator].weight + validator.weight == weight,
+            "PoSValidatorManager: Invalid weight"
+        );
+        require(validator.messageNonce >= nonce, "PoSValidatorManager: Invalid nonce");
+        validator.weight = weight;
+        $._delegatorStakes[validationID][delegator].status = ValidatorStatus.Active;
+        _setValidator(validationID, validator);
+
+        emit DelegatorRegistered(
+            validationID,
+            delegator,
+            $._delegatorStakes[validationID][delegator].weight,
+            weight,
+            block.timestamp
+        );
+    }
+
+    function initializeEndDelegation(bytes32 validationID) external {
+        PoSValidatorManagerStorage storage $ = _getPoSValidatorManagerStorage();
+
+        // Ensure the delegator is active
+        Delegator memory delegator = $._delegatorStakes[validationID][_msgSender()];
+        require(
+            delegator.status == ValidatorStatus.Active, "ValidatorManager: Delegator not active"
+        );
+        delegator.status = ValidatorStatus.PendingRemoved;
+        delegator.endedAt = uint64(block.timestamp);
+
+        $._delegatorStakes[validationID][_msgSender()] = delegator;
+
+        Validator memory validator = _getValidator(validationID);
+        require(validator.weight >= delegator.weight, "PoSValidatorManager: Invalid weight");
```

Comment:
thinking if this should be `>`. Doesn't seem like a validator with zero weight would make sense.